subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c

```diff
@@ -0,0 +1,743 @@
+			if (ret == 0) {
+				instance->update_state = UPD_STATE_INITIAL;
+				instance->update_result = UPD_RES_INITIAL;
+			}
+			break;
+		default:
+			ret = -EINVAL;
+			break;
+		}
+		break;
+	default:
+		ret = -EINVAL;
+		break;
+	}
+
+	return ret;
+}
+
+static int install_cb(uint16_t obj_inst_id)
+{
+	struct lwm2m_swmgmt_data *instance = NULL;
+	instance = find_index(obj_inst_id);
+
+
+	return handle_event(instance, EVENT_INSTALL);
+}
+
+int lwm2m_swmgmt_install_completed(uint16_t obj_inst_id, int error_code)
+{
+	struct lwm2m_swmgmt_data *instance = NULL;
+	instance = find_index(obj_inst_id);
+
+
+	if (error_code) {
+		error_code = handle_event(instance, EVENT_INSTALL_FAIL);
+	} else {
+		error_code = handle_event(instance, EVENT_INSTALL_SUCCESSFUL);
+	}
+
+	return error_code;
+}
+
+static int uninstall_cb(uint16_t obj_inst_id)
+{
+	/*
+	 * TODO Implement handling for "ForUpdate".
+	 * This requires the Zephyr LWM2M to handle arguments to exec calls.
+         */
+
+	struct lwm2m_swmgmt_data *instance = NULL;
+	instance = find_index(obj_inst_id);
+
+	return handle_event(instance, EVENT_DELETE_PACKAGE);
+}
+
+static int activate_cb(uint16_t obj_inst_id)
+{
+	struct lwm2m_swmgmt_data *instance = NULL;
+	instance = find_index(obj_inst_id);
+
+	return handle_event(instance, EVENT_ACTIVATE);
+}
+
+static int deactivate_cb(uint16_t obj_inst_id)
+{
+	struct lwm2m_swmgmt_data *instance = NULL;
+	instance = find_index(obj_inst_id);
+
+	return handle_event(instance, EVENT_DEACTIVATE);
+}
+
+static int package_write_cb(uint16_t obj_inst_id, uint16_t res_id, uint16_t res_inst_id,
+			    uint8_t *data, uint16_t data_len, bool last_block, size_t total_size)
+{
+	int ret = -EINVAL;
+	struct lwm2m_swmgmt_data *instance = NULL;
+	instance = find_index(obj_inst_id);
+
+
+	ret = handle_event(instance, EVENT_PKG_URI_WRITE);
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = instance->write_package_cb(obj_inst_id, res_id, res_inst_id, data, data_len,
+					 last_block, total_size);
+
+	if (ret < 0) {
+		handle_event(instance, EVENT_DOWNLOAD_FAILED);
+		switch (ret) {
+		case -ENOMEM:
+			instance->update_result = UPD_RES_OUT_OF_MEMORY_DURING_DOWNLOAD;
+			break;
+		case -ENOSPC:
+			instance->update_result = UPD_RES_NOT_ENOUGH_STORAGE;
+			ret = -EFBIG;
+			break;
+		case -EFAULT:
+			instance->update_result = UPD_RES_PACKAGE_INTEGRITY_CHECK_FAILURE;
+			break;
+		default:
+			instance->update_result = UPD_RES_LOST_CONNECTION_DURING_DOWNLOAD;
+			break;
+		}
+
+		return ret;
+	}
+
+	if (last_block) {
+		handle_event(instance, EVENT_PKG_WRITTEN);
+	}
+
+	return 0;
+}
+
+static void set_update_result_from_error(struct firmware_pull_context *context, int error_code)
+{
+	struct lwm2m_swmgmt_data *instance;
+	instance = CONTAINER_OF(context, struct lwm2m_swmgmt_data, pull_context);
+
+
+	if (error_code == 0) {
+		handle_event(instance, EVENT_PKG_WRITTEN);
+
+		/* Do this somehow? */
+		handle_event(instance, EVENT_PKG_INTEGRITY_VERIFIED);
```

Comment:
There should be some possibility to have an external function doing the package integrity verification and reporting the results to the object. Currently I can't see a clear path leading here.

Reply:
When you say `external` do you mean something settable by the Zephyr's user? A callback that they can set or should it be a static function in the lwm2m scope? Is it fine if we leave that for a future PR or do you think this has to be a part of this PR?

I'd obviously like it to do this but this PR is already pretty big. The comment needs fixing for sure if we choose to implement this later.

Reply:
I've implemented this but I did per download request. Do you see any value in storing a verify function per object instance?